spec/index.bs

```diff
@@ -1493,6 +1497,103 @@ interface GPUTextureUsage {
+      1. Return one plus the greatest integral value of |x| for which 2^|x| <= |m|.
+
+    </dfn>
+  </div>
+
+  <div algorithm class=validusage>
+    <dfn abstract-op>createTexture Valid Usage</dfn>
+      Given a {{GPUDevice}} |this| and a {{GPUTextureDescriptor}} |descriptor|
+      the following validation rules apply:
+
+      1. |this| must be a [=valid=] {{GPUDevice}}.
+      1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/width=] must be nonzero.
+      1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/height=] must be nonzero.
+      1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/depth=] must be nonzero.
+      1. |descriptor|.{{GPUTextureDescriptor/mipLevelCount}} must be nonzero.
+      1. |descriptor|.{{GPUTextureDescriptor/sampleCount}} must be nonzero.
+      1. If |descriptor|.{{GPUTextureDescriptor/dimension}} is "1d":
+
+        1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/width=] must be less than or equal to 8192.
+        1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/height=] must be less than or equal to 2048.
+        1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/depth=] must be 1.
+        1. |descriptor|.{{GPUTextureDescriptor/sampleCount}} must be 1.
+
+      1. Else if |descriptor|.{{GPUTextureDescriptor/dimension}} is "2d":
+
+        1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/width=] must be less than or equal to 8192.
+        1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/height=] must be less than or equal to 8192.
+        1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/depth=] must be less than or equal to 2048.
+
+      1. Else (|descriptor|.{{GPUTextureDescriptor/dimension}} is "3d"):
+
+        1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/width=] must be less than or equal to 2048.
+        1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/height=] must be less than or equal to 2048.
+        1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/depth=] must be less than or equal to 2048.
+        1. |descriptor|.{{GPUTextureDescriptor/sampleCount}} must be 1.
+
+      1. If |descriptor|.{{GPUTextureDescriptor/sampleCount}} > 1:
+
+        1. |descriptor|.{{GPUTextureDescriptor/mipLevelCount}} must be 1.
+        1. |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/depth=] must be 1.
+
+      1. |descriptor|.{{GPUTextureDescriptor/mipLevelCount}} must be less than or equal to the [$maximum valid mipLevel$] for its
+         |descriptor|.{{GPUTextureDescriptor/dimension}} and |descriptor|.{{GPUTextureDescriptor/size}}.
+
+      1. If |descriptor|.{{GPUTextureDescriptor/format}} is a [=depth or stencil format=]:
+
+        1. |descriptor|.{{GPUTextureDescriptor/dimension}} must be "2d".
+        1. |descriptor|.{{GPUTextureDescriptor/sampleCount}} must be 1.
+
+      1. |descriptor|.{{GPUTextureDescriptor/usage}} must be a combination of {{GPUTextureUsage}} values.
+
```

Comment:
There is some missing validation:

 - first mip level width and height is a multiple of the compressed block size.
 - compressed textures can only be copySrc / copyDst / Sampled.
 - other usage constraints (we have non-renderable types), or add an issue for it. Dawn's [Format.cpp](https://source.chromium.org/chromium/chromium/src/+/master:third_party/dawn/src/dawn_native/Format.cpp;l=182) summarizes what we have found by investigating the various APIs, and all usages of all formats are tested.

Reply:
Unextended WebGPU doesn't have any compressed textures, and we haven't added any extensions to add any yet, and our implementation doesn't support compressed textures yet, so I'm not comfortable adding rules for compressed textures. I can open an issue for it.

I'll open a new issue for usage constraints. AFAICT Metal doesn't have any constraints here, so I'm not sure what to restrict.

Reply:
https://github.com/gpuweb/gpuweb/issues/807
https://github.com/gpuweb/gpuweb/issues/808

Reply:
`"texture-compression-bc"` is already there although the formats enums aren't in the spec yet. We're pretty comfortable adding text for compressed textures because @Jiawei-Shao did extensive investigation and testing in Dawn so we know they work in all APIs with the current validation constraints.

There's a bit a of chicken and egg issues because you've voiced concern about adding the extension because we didn't have validation for it, and now you have concerns adding the validation because we don't have the extension yet ^^.

That said we can do compressed texture validation in a follow-up if it feels more comfortable.

Metal does have a large number of usage constraints on texture that you can find starting page 6 of the [Metal Feature Set document](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf). We can add that independently since it will likely be a large PR in itself.

Reply:
BTW though there are no compressed textures in the spec, there is a lot of setup for them (block based formats/texel blocks). You can rely on that ("block size > 1") to talk about compressed formats.

In this case: In practice this should be correct, but if we want we could eventually add an explicit flag to a texture format capability table that says "supports multiple samples".